packages/platform-core/src/aws_client_provider.ts

```diff
@@ -0,0 +1,82 @@
+import { AwsCredentialIdentityProvider } from '@aws-sdk/types';
+import { S3Client } from '@aws-sdk/client-s3';
+import { AmplifyClient } from '@aws-sdk/client-amplify';
+import { CloudFormationClient } from '@aws-sdk/client-cloudformation';
+import { fromNodeProviderChain } from '@aws-sdk/credential-providers';
+
+export type AWSClientsOverrideClientOptions = {
+  s3Client: S3Client;
+  amplifyClient: AmplifyClient;
+  cloudformationClient: CloudFormationClient;
+};
+
+export type AWSClientsCredentialsOptions = {
+  credentials: AwsCredentialIdentityProvider;
+};
+
+export type AWSClientsOptions =
+  | AWSClientsOverrideClientOptions
+  | AWSClientsCredentialsOptions;
+
+/**
+ * Instantiates AWSClientProvider
+ */
+export class AWSClientProvider {
```

Comment:
This provider needs to be generic enough that it can be used for any number of clients without pulling a bunch of dependencies for unrelated clients. I would suggest something like this:

Have a base type of
```ts
export type AWSClientProvider<
  T extends Record<`get${string}Client`, Client<object, MetadataBearer, unknown>> // these interfaces are coming from `import { Client, MetadataBearer } from '@smithy/types';`
> = {
  [K in keyof T]: () => T[K];
};
```
(Now that this is a pure type and not a class it can also be moved to `platform-types`)

Then methods/ctors/functions that need clients can specify which ones they need as a simple record:
```ts
export type GenerateModelsClientProvider = AWSClientProvider<{
  getS3Client: S3Client;
  getCfnClient: CloudFormationClient;
}>

// just an example
export const generateModels = async (
  awsClientProvider: GenerateModelsClientProvider
) => { /* do stuff with the clients */ };
```
Also note that the S3Client and CloudFormationClient dependencies can now be peerDeps rather than runtime deps of the package.

Then finally, the caller of `generateModels` can construct the client provider when calling the function

```ts
await generateModels({
  getS3Client: () => new S3Client(/* pass in custom client config */),
  getCfnClient: () => new CloudFormationClient(/* pass in custom client config */),
})
```

Reply:
The main downside in my opinion is that those peer dependencies are going to added all the way down the chain. But let me try it, maybe it won't be so bad